### PR TITLE
Add more tests to IsolatedStorage

### PR DESCRIPTION
**Pull request by JeremyKuhne**
- Adds basic coverage for most APIs
- Fixes a few issues
- Tweaks json/build files to prepare for 4.6x runs

@ianhays @ericstj @joperezr

### CODE REVIEW

File: src/System.IO.IsolatedStorage/System.IO.IsolatedStorage.sln [hunk 1]
```diff
@@ -18,6 +18,9 @@ EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.IO.IsolatedStorage", "ref\System.IO.IsolatedStorage.csproj", "{27225772-FE8B-49D7-8E58-29242D536130}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "tests", "tests", "{8751BC6B-FD4E-40E0-A5F8-63F92D987719}"
+	ProjectSection(SolutionItems) = preProject
+		tests\System.IO.IsolatedStorage.Tests.builds = tests\System.IO.IsolatedStorage.Tests.builds
```
Comment (joperezr) [inline src/System.IO.IsolatedStorage/System.IO.IsolatedStorage.sln]: I don't think we usually add the .builds files to the solution. I mean it doesn't hurt I guess, but it's just inconsistent with the rest of the repo.

File: src/System.IO.IsolatedStorage/tests/project.json [hunk 2]
```diff
@@ -32,9 +34,19 @@
     "Microsoft.DotNet.xunit.performance": "1.0.0-alpha-build0040"
   },
   "frameworks": {
-    "netstandard1.7": {}
+    "netstandard1.7": {
+      "imports": [
+        "dotnet5.8"
+      ]
+    },
+    "net463": {}
```
Comment (joperezr) [inline src/System.IO.IsolatedStorage/tests/project.json]: I thought you needed the targetting pack in order for net463 to work, is that not the case? I guess perhaps the other dependencies are pulling that in.

File: src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.builds [hunk 3]
```diff
@@ -3,15 +3,9 @@
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <ItemGroup>
     <Project Include="System.IO.IsolatedStorage.Tests.csproj">
-      <OSGroup>Windows_NT</OSGroup>
```
Comment (joperezr) [inline src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.builds]: Even though this really doesn't do anything anymore since the build doesn't depend on OSGroup, @weshaggard made the comment that we should probably keep the OSGroup when we are testing Desktop just to be consistent with the rest of the repo

Comment (weshaggard) [inline src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.builds]: I'd be interested in @ericstj thoughts on this but yes I would prefer we keep it for consistency right now and we can decide to fold them all later if needed.

Also with the current settings in this .builds file (i.e. no OSGroup here) you will have a bin-clash between the 2 configurations, as the build options (OSGroup/TargetGroup) are the same for 2 of these configurations. The TestTFMs is not part of the build configuration so don't get factored into the output directories of the build, thus we cannot have configurations that only differ by TestTFMs.

File: src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.csproj [hunk 4]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" DefaultTargets="Build">
   <PropertyGroup>
```
Comment (weshaggard) [inline src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.csproj]: This TestTFM property isn't needed.

Comment (weshaggard) [inline src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.csproj]: Also the RuntimeTargetMonikers aren't needed below.

File: src/System.IO.IsolatedStorage/tests/project.json [hunk 5]
```diff
@@ -32,9 +34,19 @@
...
   },
   "supports": {
-    "coreFx.Test.netcoreapp1.1": {}
+    "coreFx.Test.netcoreapp1.1": {},
+    "coreFx.Test.net463": {}
+  },
+  "runtimes": {
```
Comment (weshaggard) [inline src/System.IO.IsolatedStorage/tests/project.json]: Runtimes aren't needed.

File: src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.Win32.Unix.cs [hunk 6]
```diff
@@ -68,6 +68,11 @@ internal static void GetDefaultIdentityAndHash(out object identity, out string h
             // by pulling directly from EntryAssembly.
 
             Assembly assembly = Assembly.GetEntryAssembly();
+
+            // Fall back on the executing assembly if needed
```
Comment (ericstj) [inline src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.Win32.Unix.cs]: How is this ever correct?  Won't this be the "System.IO.IsolatedStorage, ..."?  Maybe we discussed this before but I can't remember the reasoning which means we probably need a comment explaining.

File: src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/IsolatedStorage.cs [hunk 7]
```diff
@@ -159,7 +159,7 @@ protected void InitStore(IsolatedStorageScope scope, Type domainEvidenceType, Ty
                     hash = $"{hash}{SeparatorExternal}{hash}";
                 }
 
-                _applicationIdentity = identity;
+                _assemblyIdentity = identity;
```
Comment (ericstj) [inline src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/IsolatedStorage.cs]: Just a typo?

File: src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.csproj [hunk 8]
```diff
@@ -1,4 +1,4 @@
...
     <TestTFM Condition="'$(TestTFM)' == ''">netcoreapp1.1</TestTFM>
```
Comment (ericstj) [inline src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.csproj]: All the metadata on the pkgproj reference below can be cleaned up.

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        public void GetCreationTime_GetsTime()
+        {
+            using (IsolatedStorageFile isf = IsolatedStorageFile.GetUserStoreForAssembly())
+            {
+                string file = "GetCreationTime_GetsTime";
+                DateTimeOffset currentTime = DateTimeOffset.Now;
+                isf.CreateTestFile(file);
+                DateTimeOffset fileTime = isf.GetCreationTime(file);
+                TimeSpan difference = fileTime - currentTime;
+                Assert.True(difference.TotalSeconds < 1.0);
```
Comment (ericstj) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs]: I've seen these types of tests fail before.  Is there any reason to make this assertion?

Comment (ericstj) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs]: Maybe a better test case would be to take the time after it is created and then check that the creation time is between before creation and after creation.

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetLastAccessTimeTests.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        public void GetLastAccessTime_GetsTime()
+        {
+            using (IsolatedStorageFile isf = IsolatedStorageFile.GetUserStoreForAssembly())
+            {
+                string file = "GetLastAccessTime_GetsTime";
+                DateTimeOffset currentTime = DateTimeOffset.Now;
+                isf.CreateTestFile(file);
+                DateTimeOffset fileTime = isf.GetLastAccessTime(file);
+                TimeSpan difference = fileTime - currentTime;
+                Assert.True(difference.TotalSeconds < 1.0);
```
Comment (ericstj) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetLastAccessTimeTests.cs]: Same as above.

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetLastWriteTimeTests.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        public void GetLastWriteTime_GetsTime()
+        {
+            using (IsolatedStorageFile isf = IsolatedStorageFile.GetUserStoreForAssembly())
+            {
+                string file = "GetLastWriteTime_GetsTime";
+                DateTimeOffset currentTime = DateTimeOffset.Now;
+                isf.CreateTestFile(file);
+                DateTimeOffset fileTime = isf.GetLastWriteTime(file);
+                TimeSpan difference = fileTime - currentTime;
+                Assert.True(difference.TotalSeconds < 1.0);
```
Comment (ericstj) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetLastWriteTimeTests.cs]: Same as above

File: src/System.IO.IsolatedStorage/tests/project.json [hunk 12]
```diff
@@ -32,9 +34,19 @@
     "Microsoft.DotNet.xunit.performance": "1.0.0-alpha-build0040"
   },
   "frameworks": {
-    "netstandard1.7": {}
+    "netstandard1.7": {
+      "imports": [
```
Comment (ericstj) [inline src/System.IO.IsolatedStorage/tests/project.json]: Why did you add this?  Imports should never be needed for our tests since they should be picking up the latest packages which all support netstandard.

File: src/System.IO.IsolatedStorage/tests/project.json (hunk 2 above)
Comment (ericstj) [inline src/System.IO.IsolatedStorage/tests/project.json]: Targeting pack is only needed if you want to compile against mscorlib-based surface.  I test can compile against System.Runtime-based surface using packages then just run on desktop.

File: src/System.IO.IsolatedStorage/tests/project.json (hunk 12 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/tests/project.json]: Accidental- copying or merging I guess.

File: src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.Win32.Unix.cs (hunk 6 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.Win32.Unix.cs]: Any ideas on what to do here instead? This happened when I tried to run against desktop.

File: src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/IsolatedStorage.cs (hunk 7 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/IsolatedStorage.cs]: Yes, stupid typo bug on my part. Once I wrote the tests I caught it.

File: src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.builds (hunk 3 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.builds]: Ok, I'll roll it back for now. It seems less than optimal to have three separate builds that all build the same binary. In addition we're really building AnyCPU, which makes this pretty confusing. It would make a lot more sense if the TestTFM figured into the actual test directory IMHO.

File: src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.csproj (hunk 8 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/tests/System.IO.IsolatedStorage.Tests.csproj]: Ok, thanks.

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/OpenFileTests.cs [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+                using (isf.OpenFile(file, FileMode.CreateNew)) { }
+                Assert.True(isf.FileExists(file), "file exists");
+
+                Assert.Throws<IsolatedStorageException>(() => isf.OpenFile(file, FileMode.CreateNew));
+                using (isf.OpenFile(file, FileMode.Create)) { }
+            }
+        }
+
+        [Theory MemberData(nameof(ValidStores))]
+        public void OpenFile_Existance(PresetScopes scope)
```
Comment (ianhays) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/OpenFileTests.cs]: nit: OpenFile_Existence

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/CreateFileTests.cs [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                Assert.Throws<InvalidOperationException>(() => isf.CreateFile("foo"));
+            }
+        }
+
+        [Fact]
+        public void CreateFile_RaisesArgumentException()
+        {
+            using (IsolatedStorageFile isf = IsolatedStorageFile.GetUserStoreForAssembly())
+            {
+                Assert.Throws<ArgumentException>(() => isf.CreateFile("\0bad"));
```
Comment (ianhays) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/CreateFileTests.cs]: Is the null character the only one we should be checking here? What about other invalid chars?

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs (hunk 9 above)
Comment (ianhays) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs]: >Maybe a better test case would be to take the time after it is created and then check that the creation time is between before creation and after creation.

👍 these tests shouldn't ever assert that they'll run in any timely fashion.

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/CreateFileTests.cs [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+
+namespace System.IO.IsolatedStorage
+{
+    public class CreateFileTests : IsoStorageTest
+    {
+        [Fact]
+        public void CreateFile_ThrowsArugmentNull()
```
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/CreateFileTests.cs]: Whoops, thanks

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/CreateFileTests.cs (hunk 14 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/CreateFileTests.cs]: Null is the only one that doesn't vary per platform. This is more about raising the exception, not completeness of the underlying functionality.

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs (hunk 9 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs]: @ericstj, @ianhays What they are trying to is allow for granularity differences in the filesystem and see that we're getting something remotely correct. I can't check times before and after as that will likely fail due to granularity. I'll make the span much greater and add a comment. I'm definitely *not* trying to check for timeliness.

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/OpenFileTests.cs (hunk 13 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/OpenFileTests.cs]: Thanks

**Comment by JeremyKuhne** [comment]
Addressed all of the feedback

**Comment by JeremyKuhne** [comment]
Made a minor fix to GetFileNames validation to allow order to vary

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs (hunk 9 above)
Comment (ericstj) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs]: >  can't check times before and after as that will likely fail due to granularity.

How so?  Isn't this acceptable
before <= filetime <= now?

File: src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.Win32.Unix.cs (hunk 6 above)
Comment (ericstj) [inline src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.Win32.Unix.cs]: I guess cross compile for desktop is one answer.

File: src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs (hunk 9 above)
Comment (JeremyKuhne) [inline src/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/GetCreationTimeTests.cs]: Because the file times vary in granularity, caching, etc. https://msdn.microsoft.com/en-us/library/windows/desktop/ms724290(v=vs.85).aspx

The system clock is way more precise. I don't want to have to allow for all the various roundings involved across all of the file systems involved. I may have to tweak this further as it is- FAT access time is rounded to the day for example.

**Comment by JeremyKuhne** [comment]
@ericstj I think I've addressed your feedback. If there is anything else I'll follow up in my next change.
